ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -122,58 +145,81 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
-      for (BlobStore store : stores) {
-        if (store.isStarted()) {
-          try {
-            store.maybeResumeCompaction();
-          } catch (Exception e) {
-            logger.error("Compaction of store {} failed on resume. Continuing with the next store", store, e);
-            storesToSkip.add(store);
-          }
-        }
-      }
-      // continue to do compactions as required.
-      while (enabled) {
-        try {
-          long startTimeMs = time.milliseconds();
-          for (BlobStore store : stores) {
+      isRunning = true;
+      startLatch.countDown();
+      try {
+        logger.info("Starting compaction thread");
+        // complete any compactions in progress
+        for (BlobStore store : stores) {
+          logger.trace("{} being checked for resume", store);
+          if (store.isStarted()) {
+            logger.trace("{} is started and eligible for resume check", store);
             try {
-              if (!enabled) {
-                break;
-              }
-              if (store.isStarted() && !storesToSkip.contains(store)) {
-                CompactionDetails details = getCompactionDetails(store);
-                if (details != null) {
-                  store.compact(details);
-                }
-              }
+              metrics.markCompactionStart(false);
+              store.maybeResumeCompaction();
+              metrics.markCompactionStop();
             } catch (Exception e) {
-              logger.error("Compaction of store {} failed. Continuing with the next store", store, e);
+              metrics.compactionErrorCount.inc();
+              logger.error("Compaction of store {} failed on resume. Continuing with the next store", store, e);
               storesToSkip.add(store);
             }
           }
-          lock.lock();
+        }
+        // continue to do compactions as required.
+        while (enabled) {
           try {
-            if (enabled) {
-              long timeElapsed = time.milliseconds() - startTimeMs;
-              time.await(waitCondition, waitTimeMs - timeElapsed);
+            long startTimeMs = time.milliseconds();
+            for (BlobStore store : stores) {
+              logger.trace("{} being checked for compaction", store);
+              try {
+                if (!enabled) {
+                  logger.trace("Breaking out because compaction thread is disabled");
+                  break;
+                }
+                if (store.isStarted() && !storesToSkip.contains(store)) {
+                  logger.trace("{} is started and eligible for compaction check", store);
+                  CompactionDetails details = getCompactionDetails(store);
+                  if (details != null) {
+                    logger.trace("Generated {} as details for {}", details, store);
+                    metrics.markCompactionStart(true);
+                    store.compact(details);
+                    metrics.markCompactionStop();
```

Comment:
please double check that marking the end of things are being done in the right place (that is, do it in the `finally` if you want to cover all exit paths).